url6f34cc586e_mekku93_sxl_interpreter_repl_tgz-pJ8-com_miguel_sxl_SXLJ8/src/com/miguel/sxl/SXLVistor.java

```diff
@@ -0,0 +1,923 @@
+	
+	/**
+	 * TYPE CAST
+	 */
+	@Override
+	public Object visit(ASTTypeCast node, Object data) {
+		if ( node.jjtGetNumChildren() == 0 ) return null;
+		
+		try {
+			// Visit the first node - the type
+			Object typeObj = node.jjtGetChild(0).jjtAccept(this, data);
+			if ( typeObj == null || !(typeObj instanceof SXLValue) ) {
+				throw new Exception("Invalid type in type cast" );
+			}
+		
+			// Get the second node - the value to cast
+			Object valueObj = node.jjtGetChild(1).jjtAccept(this, data);
+			if ( valueObj == null || !(valueObj instanceof SXLValue) ) {
+				throw new Exception("Invalid expression value in type cast");
+			}
+			
+			// Get the type
+			String type = ((SXLValue)typeObj).getType();
+			// Get the value
+			SXLValue value = (SXLValue)valueObj;
+			
+			return value.castInto(type);
+		}
+		catch( Exception e ) {
+			System.err.println(e.getMessage() + ", " + node.getPosition());
+			return null;
+		}
+	}
+
+
+
+	
+	
+	
+	/**
+	 * UNARY
+	 */
+	@Override
+	public Object visit(ASTUnary node, Object data) {
+		try {
+			// Get the unary operator
+			String op = (String)node.jjtGetChild(0).jjtAccept(this, data);
+
+			// Get the second node - the value to perform unary on
+			Object operand = node.jjtGetChild(1).jjtAccept(this, data);
+			if ( operand == null || !(operand instanceof SXLValue) ) {
+				throw new Exception("Invalid expression value operand for unary operation");
+			}
+			
+			// Get the value
+			SXLValue value = (SXLValue)operand;
+			// Get the unary result
+			SXLValue result = value.unary(op);
+			
+			// Compute the result of the unary operation and return it
+			return Operation.perform(op, (SXLValue)value, null);
+		} catch ( Exception e ) {
+			System.err.println(e.getMessage() + ", " + node.getPosition());
+			return null;
+		}
+	}
+
+	
+	
+	
+	
+	/**
+	 * FACTOR
+	 */
+	@Override
+	public Object visit(ASTFactor node, Object data) {
+		if ( node.jjtGetNumChildren() == 0 ) return null;
+		
+		// visit the first node
+		return node.jjtGetChild(0).jjtAccept(this, data);
+	}
+
+	/**
+	 * TERM
+	 */
+	@Override
+	public Object visit(ASTTerm node, Object data) {
+		// If no nodes
+		if ( node.children.length == 0 ) return null;
+		
+		try {
+			// Visit first node and get data
+			Object obj1 = node.children[0].jjtAccept(this, data);
+			
+			// If this is the only node, return its value
+			if ( node.jjtGetNumChildren() == 1 ) {
+				return obj1;
+			}
+			// Otherwise, visit the operator and the last node
+			else {
+				// Visit operator and third node and get their data
+				Object operator = node.jjtGetChild(1).jjtAccept(this, data);
+				Object obj2 = node.jjtGetChild(2).jjtAccept(this, data);
+				
+				// Perform the operation. Can only be multiplicative due to AST structure
+				Object value = Operation.perform(operator, (SXLValue)obj1, (SXLValue)obj2);
+				return value;
+			}
+		}
+		catch( Exception e ) {
+			System.err.println(e.getMessage() + ", " + node.getPosition());
+			return null;
+		}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 353:377, 387:410.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.